x-pack/plugins/maps/common/descriptor_types/layer_descriptor_types.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+/* eslint-disable @typescript-eslint/consistent-type-definitions */
+
+import { Query } from 'src/plugins/data/public';
+import { StyleDescriptor, VectorStyleDescriptor } from './style_property_descriptor_types';
+import { DataRequestDescriptor } from './data_request_descriptor_types';
+import { AbstractSourceDescriptor, ESTermSourceDescriptor } from './source_descriptor_types';
+
+export type JoinDescriptor = {
+  leftField?: string;
+  right: ESTermSourceDescriptor;
+};
+
+export type LayerDescriptor = {
+  __dataRequests?: DataRequestDescriptor[];
+  __isInErrorState?: boolean;
+  __isPreviewLayer?: boolean;
+  __errorMessage?: string;
+  __trackedLayerDescriptor?: LayerDescriptor;
+  alpha?: number;
+  id: string;
+  joins?: JoinDescriptor[];
+  label?: string | null;
+  areLabelsOnTop?: boolean;
+  minZoom?: number;
+  maxZoom?: number;
+  sourceDescriptor: AbstractSourceDescriptor | null;
```

Comment:
Do you remember how a `null` sourceDescriptor can arise?

Reply:
I do not